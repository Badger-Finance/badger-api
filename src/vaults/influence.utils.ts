import { formatBalance, Vault__factory } from '@badger-dao/sdk';

import { Chain } from '../chains/config/chain.config';
import { TOKENS } from '../config/tokens.config';
import { getBalancerPoolTokens } from '../protocols/strategies/balancer.strategy';

// TODO: setup influence configs, voting periods, etc.
const influenceVaults = new Set([TOKENS.BVECVX, TOKENS.GRAVI_AURA]);

export function isInfluenceVault(address: string) {
  return influenceVaults.has(address);
}

export async function getInfuelnceVaultYieldBalance(
  chain: Chain,
  address: string,
  targetBlock: number,
): Promise<number> {
  const sdk = await chain.getSdk();
  const vaultContract = Vault__factory.connect(address, sdk.provider);
  const strategyBalance = await vaultContract.totalSupply({ blockTag: targetBlock });

  const maxBalance = formatBalance(strategyBalance);
  // TODO: tidy this up probably, or rethink with the configs how we deal with this
  if (address === TOKENS.BVECVX) {
    return maxBalance;
  }

  // deal with GRAVI_AURA - this won't scale so we need to figure out how to generalize it
  const excludedPools = [TOKENS.BPT_GRAV_AURABAL_WETH, TOKENS.BPT_GRAV_DIGG_WBTC];

  let blacklistedBalance = 0;
  for (const pool of excludedPools) {
    const tokens = await getBalancerPoolTokens(chain, pool);
    const graviAura = tokens.find((t) => t.address === TOKENS.GRAVI_AURA);
    if (graviAura) {
      blacklistedBalance += graviAura.balance;
    }
  }

  return maxBalance - blacklistedBalance;
}
