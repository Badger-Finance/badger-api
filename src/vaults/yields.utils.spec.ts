import { ONE_DAY_MS, VaultState } from '@badger-dao/sdk';

import { TOKENS } from '../config/tokens.config';
import { SourceType } from '../rewards/enums/source-type.enum';
import { MOCK_VAULT, MOCK_VAULT_DEFINITION } from '../test/constants';
import { mockBalance, mockQuery } from '../test/mocks.utils';
import { fullTokenMockMap } from '../tokens/mocks/full-token.mock';
import { VAULT_SOURCE } from './vaults.utils';
import {
  calculateBalanceDifference,
  calculateYield,
  createYieldSource,
  getVaultYieldProjection,
  getYieldSources,
} from './yields.utils';

describe('yields.utils', () => {
  const baseMockSources = [
    createYieldSource(MOCK_VAULT_DEFINITION, SourceType.PreCompound, VAULT_SOURCE, 7),
    createYieldSource(MOCK_VAULT_DEFINITION, SourceType.Compound, VAULT_SOURCE, 11),
    createYieldSource(MOCK_VAULT_DEFINITION, SourceType.TradeFee, 'LP Fees', 2),
    createYieldSource(MOCK_VAULT_DEFINITION, SourceType.Emission, 'Badger', 4),
    createYieldSource(MOCK_VAULT_DEFINITION, SourceType.Emission, 'Boosted Badger', 6, { min: 0.5, max: 2 }),
    createYieldSource(MOCK_VAULT_DEFINITION, SourceType.Flywheel, 'Vault Flywheel', 5),
    createYieldSource(MOCK_VAULT_DEFINITION, SourceType.Distribution, 'Badger', 3),
    createYieldSource(MOCK_VAULT_DEFINITION, SourceType.Distribution, 'Irrelevant', 0.0001),
  ];

  describe('calculateYield', () => {
    it.each([
      [365, 1, ONE_DAY_MS, 0, 100],
      [365, 0.5, ONE_DAY_MS, 0, 50],
      [365, 2, ONE_DAY_MS, 0, 200],
      [365, 1, ONE_DAY_MS, 1, 171.45674820219733],
      [365, 1, ONE_DAY_MS, 0.5, 114.81572517391494],
      [0, 1, ONE_DAY_MS, 0, 0],
    ])(
      '%d earned %d over %d ms with %d compounded, for %d apr',
      (principal, earned, duration, compoundingValue, expected) => {
        expect(calculateYield(principal, earned, duration, compoundingValue)).toEqual(expected);
      },
    );

    it('throws an error when provided with invalid principal and compounding pair', () => {
      expect(() => calculateYield(365, 1, ONE_DAY_MS, 2)).toThrow(
        'Compounding value must be less than or equal to earned',
      );
    });
  });

  describe('calculateBalanceDifference', () => {
    it('returns an array with the difference in token amounts', () => {
      const badger = fullTokenMockMap[TOKENS.BADGER];
      const wbtc = fullTokenMockMap[TOKENS.WBTC];
      const listA = [mockBalance(badger, 10), mockBalance(wbtc, 2)];
      const listB = [mockBalance(badger, 25), mockBalance(wbtc, 5)];
      expect(calculateBalanceDifference(listA, listB)).toMatchObject([mockBalance(badger, 15), mockBalance(wbtc, 3)]);
    });
  });

  describe('getYieldSources', () => {
    it('returns vault yield sources categorized by required breakdown', async () => {
      mockQuery(baseMockSources);
      const yieldSources = await getYieldSources(MOCK_VAULT_DEFINITION);
      expect(yieldSources).toMatchSnapshot();
    });

    it('returns only passive yield sources for a discontinued vault', async () => {
      const mockSources = [
        createYieldSource(MOCK_VAULT_DEFINITION, SourceType.PreCompound, VAULT_SOURCE, 7),
        createYieldSource(MOCK_VAULT_DEFINITION, SourceType.Compound, VAULT_SOURCE, 11),
        createYieldSource(MOCK_VAULT_DEFINITION, SourceType.TradeFee, 'LP Fees', 2),
        createYieldSource(MOCK_VAULT_DEFINITION, SourceType.Flywheel, 'Vault Flywheel', 5),
        createYieldSource(MOCK_VAULT_DEFINITION, SourceType.Distribution, 'Badger', 3),
        createYieldSource(MOCK_VAULT_DEFINITION, SourceType.Distribution, 'Irrelevant', 0.0001),
      ];
      mockQuery(mockSources);
      const definitionCopy = JSON.parse(JSON.stringify(MOCK_VAULT_DEFINITION));
      definitionCopy.state = VaultState.Discontinued;
      const yieldSources = await getYieldSources(definitionCopy);
      expect(yieldSources).toMatchSnapshot();
    });
  });

  describe('getVaultYieldProjection', () => {
    it('returns a yield projection estimation from given inputs', async () => {
      const mockDefinition = JSON.parse(JSON.stringify(MOCK_VAULT_DEFINITION));
      mockDefinition.address = MOCK_VAULT.vaultToken;
      mockQuery(baseMockSources);
      jest.spyOn(Date, 'now').mockImplementation(() => referenceTime);
      const yieldSources = await getYieldSources(mockDefinition);
      const referenceTime = 1660071985202;
      const wbtc = fullTokenMockMap[TOKENS.WBTC];
      const mockVault = JSON.parse(JSON.stringify(MOCK_VAULT));
      mockVault.lastHarvest = referenceTime - ONE_DAY_MS * 30;
      const mockYieldEstimate = {
        vault: MOCK_VAULT.vaultToken,
        yieldTokens: [mockBalance(wbtc, 0.0092)],
        harvestTokens: [mockBalance(wbtc, 0.0091)],
        previousYieldTokens: [mockBalance(wbtc, 0.0048)],
        previousHarvestTokens: [mockBalance(wbtc, 0.0047)],
        duration: ONE_DAY_MS * 15,
        lastMeasuredAt: referenceTime - ONE_DAY_MS * 15,
        lastReportedAt: 0,
        lastHarvestedAt: referenceTime - ONE_DAY_MS * 30,
      };
      const result = getVaultYieldProjection(mockVault, yieldSources, mockYieldEstimate);
      expect(result).toMatchSnapshot();
    });
  });
});
