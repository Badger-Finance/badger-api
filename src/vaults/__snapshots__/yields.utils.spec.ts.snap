// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yields.utils getVaultYieldProjection returns a yield projection estimation from given inputs 1`] = `
Object {
  "harvestApr": 27.046043378488417,
  "harvestPeriodApr": 26.15441557480198,
  "harvestPeriodApy": 26.15441557480198,
  "harvestPeriodSources": Array [
    Object {
      "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
      "apr": 26.15441557480198,
      "balance": 0.0044,
      "decimals": 8,
      "name": "Wrapped Bitcoin",
      "symbol": "WBTC",
      "value": 2.42,
    },
  ],
  "harvestPeriodSourcesApy": Array [
    Object {
      "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
      "apr": 26.15441557480198,
      "balance": 0.0044,
      "decimals": 8,
      "name": "Wrapped Bitcoin",
      "symbol": "WBTC",
      "value": 2.42,
    },
  ],
  "harvestTokens": Array [
    Object {
      "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
      "apr": 27.046043378488417,
      "balance": 0.0091,
      "decimals": 8,
      "name": "Wrapped Bitcoin",
      "symbol": "WBTC",
      "value": 5.005,
    },
  ],
  "harvestValue": 5.005,
  "nonHarvestApr": 12,
  "nonHarvestApy": 17,
  "nonHarvestSources": Array [
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": false,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_lp_fees_flat",
      "name": "LP Fees",
      "performance": Object {
        "baseYield": 2,
        "grossYield": 2,
        "maxGrossYield": 2,
        "maxYield": 2,
        "minGrossYield": 2,
        "minYield": 2,
      },
      "type": "lp_fee",
    },
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": false,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_badger_flat",
      "name": "Badger",
      "performance": Object {
        "baseYield": 4,
        "grossYield": 4,
        "maxGrossYield": 4,
        "maxYield": 4,
        "minGrossYield": 4,
        "minYield": 4,
      },
      "type": "emission",
    },
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": true,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_boosted_badger_boosted",
      "name": "Boosted Badger",
      "performance": Object {
        "baseYield": 6,
        "grossYield": 8,
        "maxGrossYield": 16,
        "maxYield": 12,
        "minGrossYield": 4,
        "minYield": 3,
      },
      "type": "emission",
    },
  ],
  "nonHarvestSourcesApy": Array [
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": false,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_lp_fees_flat",
      "name": "LP Fees",
      "performance": Object {
        "baseYield": 2,
        "grossYield": 2,
        "maxGrossYield": 2,
        "maxYield": 2,
        "minGrossYield": 2,
        "minYield": 2,
      },
      "type": "lp_fee",
    },
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": false,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_badger_flat",
      "name": "Badger",
      "performance": Object {
        "baseYield": 4,
        "grossYield": 4,
        "maxGrossYield": 4,
        "maxYield": 4,
        "minGrossYield": 4,
        "minYield": 4,
      },
      "type": "emission",
    },
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": true,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_boosted_badger_boosted",
      "name": "Boosted Badger",
      "performance": Object {
        "baseYield": 6,
        "grossYield": 8,
        "maxGrossYield": 16,
        "maxYield": 12,
        "minGrossYield": 4,
        "minYield": 3,
      },
      "type": "emission",
    },
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": false,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_flywheel_vault_flywheel_flat",
      "name": "Vault Flywheel",
      "performance": Object {
        "baseYield": 5,
        "grossYield": 5,
        "maxGrossYield": 5,
        "maxYield": 5,
        "minGrossYield": 5,
        "minYield": 5,
      },
      "type": "flywheel",
    },
  ],
  "yieldApr": 27.34325264638389,
  "yieldPeriodApr": 26.15441557480198,
  "yieldPeriodSources": Array [
    Object {
      "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
      "apr": 26.15441557480198,
      "balance": 0.0044,
      "decimals": 8,
      "name": "Wrapped Bitcoin",
      "symbol": "WBTC",
      "value": 2.42,
    },
  ],
  "yieldTokens": Array [
    Object {
      "address": "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599",
      "apr": 27.34325264638389,
      "balance": 0.0092,
      "decimals": 8,
      "name": "Wrapped Bitcoin",
      "symbol": "WBTC",
      "value": 5.06,
    },
  ],
  "yieldValue": 5.06,
}
`;

exports[`yields.utils getYieldSources returns only passive yield sources for a discontinued vault 1`] = `
Object {
  "apr": Object {
    "baseYield": 2,
    "grossYield": 2,
    "maxGrossYield": 2,
    "maxYield": 2,
    "minGrossYield": 2,
    "minYield": 2,
    "sources": Array [
      Object {
        "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "boostable": false,
        "chain": "ethereum",
        "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_lp_fees_flat",
        "name": "LP Fees",
        "performance": Object {
          "baseYield": 2,
          "grossYield": 2,
          "maxGrossYield": 2,
          "maxYield": 2,
          "minGrossYield": 2,
          "minYield": 2,
        },
        "type": "lp_fee",
      },
    ],
  },
  "apy": Object {
    "baseYield": 2,
    "grossYield": 2,
    "maxGrossYield": 2,
    "maxYield": 2,
    "minGrossYield": 2,
    "minYield": 2,
    "sources": Array [
      Object {
        "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "boostable": false,
        "chain": "ethereum",
        "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_lp_fees_flat",
        "name": "LP Fees",
        "performance": Object {
          "baseYield": 2,
          "grossYield": 2,
          "maxGrossYield": 2,
          "maxYield": 2,
          "minGrossYield": 2,
          "minYield": 2,
        },
        "type": "lp_fee",
      },
    ],
  },
  "nonHarvestSources": Array [
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": false,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_lp_fees_flat",
      "name": "LP Fees",
      "performance": Object {
        "baseYield": 2,
        "grossYield": 2,
        "maxGrossYield": 2,
        "maxYield": 2,
        "minGrossYield": 2,
        "minYield": 2,
      },
      "type": "lp_fee",
    },
  ],
  "nonHarvestSourcesApy": Array [
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": false,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_lp_fees_flat",
      "name": "LP Fees",
      "performance": Object {
        "baseYield": 2,
        "grossYield": 2,
        "maxGrossYield": 2,
        "maxYield": 2,
        "minGrossYield": 2,
        "minYield": 2,
      },
      "type": "lp_fee",
    },
  ],
}
`;

exports[`yields.utils getYieldSources returns vault yield sources categorized by required breakdown 1`] = `
Object {
  "apr": Object {
    "baseYield": 22,
    "grossYield": 24,
    "maxGrossYield": 32,
    "maxYield": 28,
    "minGrossYield": 20,
    "minYield": 19,
    "sources": Array [
      Object {
        "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "boostable": false,
        "chain": "ethereum",
        "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_precompound_vault_compounding_flat",
        "name": "Vault Compounding",
        "performance": Object {
          "baseYield": 7,
          "grossYield": 7,
          "maxGrossYield": 7,
          "maxYield": 7,
          "minGrossYield": 7,
          "minYield": 7,
        },
        "type": "precompound",
      },
      Object {
        "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "boostable": false,
        "chain": "ethereum",
        "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_lp_fees_flat",
        "name": "LP Fees",
        "performance": Object {
          "baseYield": 2,
          "grossYield": 2,
          "maxGrossYield": 2,
          "maxYield": 2,
          "minGrossYield": 2,
          "minYield": 2,
        },
        "type": "lp_fee",
      },
      Object {
        "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "boostable": false,
        "chain": "ethereum",
        "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_badger_flat",
        "name": "Badger",
        "performance": Object {
          "baseYield": 7,
          "grossYield": 4,
          "maxGrossYield": 4,
          "maxYield": 7,
          "minGrossYield": 4,
          "minYield": 7,
        },
        "type": "emission",
      },
      Object {
        "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "boostable": true,
        "chain": "ethereum",
        "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_boosted_badger_boosted",
        "name": "Boosted Badger",
        "performance": Object {
          "baseYield": 6,
          "grossYield": 8,
          "maxGrossYield": 16,
          "maxYield": 12,
          "minGrossYield": 4,
          "minYield": 3,
        },
        "type": "emission",
      },
    ],
  },
  "apy": Object {
    "baseYield": 31,
    "grossYield": 33,
    "maxGrossYield": 41,
    "maxYield": 37,
    "minGrossYield": 29,
    "minYield": 28,
    "sources": Array [
      Object {
        "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "boostable": false,
        "chain": "ethereum",
        "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_compound_vault_compounding_flat",
        "name": "Vault Compounding",
        "performance": Object {
          "baseYield": 11,
          "grossYield": 11,
          "maxGrossYield": 11,
          "maxYield": 11,
          "minGrossYield": 11,
          "minYield": 11,
        },
        "type": "compound",
      },
      Object {
        "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "boostable": false,
        "chain": "ethereum",
        "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_lp_fees_flat",
        "name": "LP Fees",
        "performance": Object {
          "baseYield": 2,
          "grossYield": 2,
          "maxGrossYield": 2,
          "maxYield": 2,
          "minGrossYield": 2,
          "minYield": 2,
        },
        "type": "lp_fee",
      },
      Object {
        "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "boostable": false,
        "chain": "ethereum",
        "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_badger_flat",
        "name": "Badger",
        "performance": Object {
          "baseYield": 7,
          "grossYield": 4,
          "maxGrossYield": 4,
          "maxYield": 7,
          "minGrossYield": 4,
          "minYield": 7,
        },
        "type": "emission",
      },
      Object {
        "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "boostable": true,
        "chain": "ethereum",
        "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_boosted_badger_boosted",
        "name": "Boosted Badger",
        "performance": Object {
          "baseYield": 6,
          "grossYield": 8,
          "maxGrossYield": 16,
          "maxYield": 12,
          "minGrossYield": 4,
          "minYield": 3,
        },
        "type": "emission",
      },
      Object {
        "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "boostable": false,
        "chain": "ethereum",
        "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
        "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_flywheel_vault_flywheel_flat",
        "name": "Vault Flywheel",
        "performance": Object {
          "baseYield": 5,
          "grossYield": 5,
          "maxGrossYield": 5,
          "maxYield": 5,
          "minGrossYield": 5,
          "minYield": 5,
        },
        "type": "flywheel",
      },
    ],
  },
  "nonHarvestSources": Array [
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": false,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_lp_fees_flat",
      "name": "LP Fees",
      "performance": Object {
        "baseYield": 2,
        "grossYield": 2,
        "maxGrossYield": 2,
        "maxYield": 2,
        "minGrossYield": 2,
        "minYield": 2,
      },
      "type": "lp_fee",
    },
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": false,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_badger_flat",
      "name": "Badger",
      "performance": Object {
        "baseYield": 4,
        "grossYield": 4,
        "maxGrossYield": 4,
        "maxYield": 4,
        "minGrossYield": 4,
        "minYield": 4,
      },
      "type": "emission",
    },
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": true,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_boosted_badger_boosted",
      "name": "Boosted Badger",
      "performance": Object {
        "baseYield": 6,
        "grossYield": 8,
        "maxGrossYield": 16,
        "maxYield": 12,
        "minGrossYield": 4,
        "minYield": 3,
      },
      "type": "emission",
    },
  ],
  "nonHarvestSourcesApy": Array [
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": false,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_lp_fee_lp_fees_flat",
      "name": "LP Fees",
      "performance": Object {
        "baseYield": 2,
        "grossYield": 2,
        "maxGrossYield": 2,
        "maxYield": 2,
        "minGrossYield": 2,
        "minYield": 2,
      },
      "type": "lp_fee",
    },
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": false,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_badger_flat",
      "name": "Badger",
      "performance": Object {
        "baseYield": 4,
        "grossYield": 4,
        "maxGrossYield": 4,
        "maxYield": 4,
        "minGrossYield": 4,
        "minYield": 4,
      },
      "type": "emission",
    },
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": true,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_emission_boosted_badger_boosted",
      "name": "Boosted Badger",
      "performance": Object {
        "baseYield": 6,
        "grossYield": 8,
        "maxGrossYield": 16,
        "maxYield": 12,
        "minGrossYield": 4,
        "minYield": 3,
      },
      "type": "emission",
    },
    Object {
      "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "boostable": false,
      "chain": "ethereum",
      "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
      "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_flywheel_vault_flywheel_flat",
      "name": "Vault Flywheel",
      "performance": Object {
        "baseYield": 5,
        "grossYield": 5,
        "maxGrossYield": 5,
        "maxYield": 5,
        "minGrossYield": 5,
        "minYield": 5,
      },
      "type": "flywheel",
    },
  ],
}
`;

exports[`yields.utils queryVaultYieldSources vault with compounding yield contains no compounding vault sources 1`] = `
Array [
  CachedYieldSource {
    "address": "0xBA485b556399123261a5F9c95d413B4f93107407",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_precompound_vault_compounding_flat",
    "name": "Vault Compounding",
    "performance": Object {
      "baseYield": 230,
      "grossYield": 230,
      "maxGrossYield": 230,
      "maxYield": 230,
      "minGrossYield": 230,
      "minYield": 230,
    },
    "type": "precompound",
  },
  CachedYieldSource {
    "address": "0xBA485b556399123261a5F9c95d413B4f93107407",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_compound_vault_compounding_flat",
    "name": "Vault Compounding",
    "performance": Object {
      "baseYield": 800.4273964093828,
      "grossYield": 800.4273964093828,
      "maxGrossYield": 800.4273964093828,
      "maxYield": 800.4273964093828,
      "minGrossYield": 800.4273964093828,
      "minYield": 800.4273964093828,
    },
    "type": "compound",
  },
  CachedYieldSource {
    "address": "0xBA485b556399123261a5F9c95d413B4f93107407",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_distribution_badger_flat",
    "name": "Badger",
    "performance": Object {
      "baseYield": 72.2,
      "grossYield": 80.22222222222223,
      "maxGrossYield": 80.22222222222223,
      "maxYield": 72.2,
      "minGrossYield": 80.22222222222223,
      "minYield": 72.2,
    },
    "type": "distribution",
  },
  CachedYieldSource {
    "address": "0xBA485b556399123261a5F9c95d413B4f93107407",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_flywheel_vault_flywheel_flat",
    "name": "Vault Flywheel",
    "performance": Object {
      "baseYield": 0.2151821909384493,
      "grossYield": 0.2151821909384493,
      "maxGrossYield": 0.2151821909384493,
      "maxYield": 0.2151821909384493,
      "minGrossYield": 0.2151821909384493,
      "minYield": 0.2151821909384493,
    },
    "type": "flywheel",
  },
]
`;

exports[`yields.utils queryVaultYieldSources vault with no compounding yield contains no compounding vault sources 1`] = `
Array [
  CachedYieldSource {
    "address": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "boostable": false,
    "chain": "ethereum",
    "chainAddress": "ethereum-0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    "id": "ethereum-0x19d97d8fa813ee2f51ad4b4e04ea08baf4dffc28_distribution_gravitationally_bound_aura_flat",
    "name": "Gravitationally Bound AURA",
    "performance": Object {
      "baseYield": 230,
      "grossYield": 255.55555555555557,
      "maxGrossYield": 255.55555555555557,
      "maxYield": 230,
      "minGrossYield": 255.55555555555557,
      "minYield": 230,
    },
    "type": "distribution",
  },
]
`;
